# gcl_illumina_qc
![](pipeline_dag.png)
## Folder Structure
```
ğŸ“ project-root/
â”œâ”€â”€ ğŸ“ gcl_illumina_qc/               # All Nextflow workflow logic
â”‚   â”œâ”€â”€ ğŸ“„ main.nf                    # Entry point for pipeline
â”‚   â”œâ”€â”€ ğŸ“„ nextflow.config            # SLURM and container config
â”‚   â”œâ”€â”€ ğŸ“„ README.md                  # Usage and documentation
â”‚   â””â”€â”€ ğŸ“ modules/                   # Individual DSL2 module processes
â”‚       â”œâ”€â”€ clumpify.nf
â”‚       â”œâ”€â”€ fastp_trim_3.nf
â”‚       â”œâ”€â”€ fastp_trim_5.nf
â”‚       â”œâ”€â”€ fastq_screen.nf
â”‚       â”œâ”€â”€ fetch_genome.nf
â”‚       â”œâ”€â”€ index_genome.nf
â”‚       â”œâ”€â”€ map_reads.nf
â”‚       â”œâ”€â”€ multiqc.nf
â”‚       â””â”€â”€ repair.nf
â”œâ”€â”€ ğŸ“ data/                          # Input FASTQ files and QC outputs
â”œâ”€â”€ ğŸ“ genome/                        # Downloaded and indexed reference
â”œâ”€â”€ ğŸ“ logs/                          # SLURM/Nextflow job logs
â””â”€â”€ ğŸ“ results/                       # Final results (e.g. MultiQC reports)
    â””â”€â”€ ğŸ“ multiqc/                   # Per-step and consolidated MultiQC outputs
```

## To Run

1. Clone repo into your directory.
2. Run the pipeline from the project root:

```bash
module load miniconda3
source activate nextflow
nextflow run gcl_illumina_qc/main.nf -resume -with-dag flowchart.dot
dot -Tpng flowchart.dot -o flowchart.png

nextflow run gcl_illumina_qc/main.nf \
    -profile standard \
    -resume \
    --reads "data/fq_raw/*.{1,2}.fq.gz" \
    --accession "GCA_042920385.1" \
	--decontam_conffile "configs/contam_db.conf" \
    --outdir "results"
```
3. Or Run in SLURM `sbatch run_qc.sbatch`
